include/wabt/type-checker.h

```diff
@@ -20,17 +20,94 @@
 #include <functional>
 #include <type_traits>
 #include <vector>
+#include <map>
 
 #include "wabt/common.h"
 #include "wabt/feature.h"
 #include "wabt/opcode.h"
 
+#include "wabt/binary-reader.h"  // For TypeMut.
+
 namespace wabt {
 
 class TypeChecker {
  public:
   using ErrorCallback = std::function<void(const char* msg)>;
 
+  enum class TypeEntryKind {
+    Func,
+    Struct,
+    Array,
+  };
+
+  struct TypeEntry {
+    explicit TypeEntry(TypeEntryKind kind, Index map_index)
+      : kind(kind), map_index(map_index) {}
+
+    TypeEntryKind kind;
+    Index map_index;
+  };
+
+  struct FuncType {
+    FuncType() = default;
+    FuncType(const TypeVector& params,
+             const TypeVector& results,
+             Index type_index)
+        : params(params), results(results), type_index(type_index) {}
+
+    TypeVector params;
+    TypeVector results;
+    Index type_index;
+  };
+
+  struct StructType {
+    StructType() = default;
+    StructType(const TypeMutVector& fields) : fields(fields) {}
+
+    TypeMutVector fields;
+  };
+
+  struct ArrayType {
+    ArrayType() = default;
+    ArrayType(TypeMut field) : field(field) {}
+
+    TypeMut field;
+  };
+
+  struct RecursiveRange {
+    Index start_index;
+    Index type_count;
+  };
```

Comment:
This is another core structure to encode `(rec ...)` constructs. It represents the range.